src/macaron/dependency_analyzer/cyclonedx.py

```diff
@@ -35,155 +46,608 @@ def deserialize_bom_json(file_path: Path) -> dict:
+        if root_bom.metadata:
+            return root_bom.metadata.component
+        return None
 
-def convert_components_to_artifacts(
-    components: Iterable[dict], root_component: dict | None = None
-) -> dict[str, DependencyInfo]:
-    """Convert CycloneDX components using internal artifact representation.
+    def get_dep_components(
+        self,
+        target_component: Component,
+        root_bom_path: Path,
+        child_bom_paths: list[Path] | None = None,
+        recursive: bool = False,
+    ) -> Iterable[CDXComponent]:
+        """Get CycloneDX components that are dependencies of the analyzed target software component.
+
+        Parameters
+        ----------
+        target_component: Component
+            The analyzed target software component.
+        root_bom_path : str
+            Path to the root bom.json file.
+        child_bom_paths: list[Path] | None
+            The list of paths to sub-project bom.json files.
+        recursive: bool
+            Set to False to get the direct dependencies only (default).
+
+        Yields
+        ------
+        CDXComponent
+            The dependencies as CycloneDX components.
+        """
+        try:
+            root_bom = deserialize_bom_json(root_bom_path)
+        except CycloneDXParserError as error:
+            logger.error(error)
+            return
 
-    Parameters
-    ----------
-    components : list[dict]
-        The dependency components.
-    root_component: dict | None
-        The root CycloneDX component.
+        if root_bom.components is None:
+            logger.error("The BOM file at %s misses components.", str(root_bom_path))
+            return
 
-    Returns
-    -------
-    dict
-        A dictionary where dependency artifacts are grouped based on "groupId:artifactId".
-    """
-    all_versions: dict[str, list[DependencyInfo]] = {}  # Stores all the versions of dependencies for debugging.
-    latest_deps: dict[str, DependencyInfo] = {}  # Stores the latest version of dependencies.
-    url_to_artifact: dict[str, set] = {}  # Used to detect artifacts that have similar repos.
-    for component in components:
+        dependencies: list[CDXDependency] = []
+
+        # Find dependencies in the root BOM file.
         try:
+            target_cdx_component = self.get_target_cdx_component(root_bom=root_bom, target_component=target_component)
+            for node in root_bom.dependencies:
+                if not isinstance(node, CDXDependency):
+                    continue
+                if recursive or (
+                    target_cdx_component and target_cdx_component.bom_ref and node.ref == target_cdx_component.bom_ref
+                ):
+                    if dep_on := node.dependencies:
+                        dependencies.extend(dep_on)
+        except AttributeError as error:
+            logger.debug(error)
+
+        # Find dependencies in child BOMs if they exist. Multi-module Java projects need this resolution.
```

Comment:
Hmmm if only Multi-module Java projects need this resolution, should we move this logic into `cyclonedx_gradle` and `cyclonedx_mvn`?